browser/ui/webui/brave_new_tab_page_refresh/top_sites_facade.cc

```diff
@@ -0,0 +1,159 @@
+
+namespace {
+
+std::vector<mojom::TopSitePtr> TopSitesFromSections(
+    const std::map<ntp_tiles::SectionType, ntp_tiles::NTPTilesVector>&
+        sections) {
+  std::vector<mojom::TopSitePtr> top_sites;
+  for (auto& tile : sections.at(ntp_tiles::SectionType::PERSONALIZED)) {
+    auto site = mojom::TopSite::New();
+    site->title = base::UTF16ToUTF8(tile.title);
+    site->url = tile.url.spec();
+    site->favicon = tile.favicon_url.spec();
+    if (site->title.empty()) {
+      site->title = site->url;
+    }
+    top_sites.push_back(std::move(site));
+  }
+  return top_sites;
+}
+
+}  // namespace
+
+TopSitesFacade::TopSitesFacade(
+    std::unique_ptr<ntp_tiles::MostVisitedSites> most_visited_sites,
+    PrefService& pref_service)
+    : most_visited_sites_(std::move(most_visited_sites)),
+      pref_service_(pref_service) {
+  CHECK(most_visited_sites_);
+
+  pref_change_registrar_.Init(&pref_service_.get());
+  pref_change_registrar_.Add(ntp_prefs::kNtpShortcutsVisible,
+                             base::BindRepeating(&TopSitesFacade::OnPrefChanged,
+                                                 weak_factory_.GetWeakPtr()));
+  pref_change_registrar_.Add(ntp_prefs::kNtpUseMostVisitedTiles,
+                             base::BindRepeating(&TopSitesFacade::OnPrefChanged,
+                                                 weak_factory_.GetWeakPtr()));
+
+  SyncMostVisitedSites();
+  most_visited_sites_->AddMostVisitedURLsObserver(
+      this, ntp_tiles::kMaxNumMostVisited);
+}
+
+TopSitesFacade::~TopSitesFacade() = default;
+
+bool TopSitesFacade::GetTopSitesVisible() {
+  return pref_service_->GetBoolean(ntp_prefs::kNtpShortcutsVisible);
+}
+
+void TopSitesFacade::SetTopSitesVisible(bool visible) {
+  pref_service_->SetBoolean(ntp_prefs::kNtpShortcutsVisible, visible);
+  most_visited_sites_->SetShortcutsVisible(visible);
+}
+
+mojom::TopSitesListKind TopSitesFacade::GetListKind() {
+  bool use_most_visited =
+      pref_service_->GetBoolean(ntp_prefs::kNtpUseMostVisitedTiles);
+  return use_most_visited ? mojom::TopSitesListKind::kMostVisited
+                          : mojom::TopSitesListKind::kCustom;
+}
+
+void TopSitesFacade::SetListKind(mojom::TopSitesListKind list_kind) {
+  bool use_most_visited = list_kind == mojom::TopSitesListKind::kMostVisited;
+  pref_service_->SetBoolean(ntp_prefs::kNtpUseMostVisitedTiles,
+                            use_most_visited);
+  most_visited_sites_->EnableCustomLinks(!use_most_visited);
+}
+
+void TopSitesFacade::GetSites(GetSitesCallback callback) {
+  std::vector<mojom::TopSitePtr> sites;
+  sites.reserve(current_sites_.size());
+  for (auto& site : current_sites_) {
+    sites.push_back(site.Clone());
+  }
+  std::move(callback).Run(std::move(sites));
+}
+
+void TopSitesFacade::AddCustomSite(std::string_view url,
+                                   std::string_view title) {
+  most_visited_sites_->AddCustomLink(GURL(url), base::UTF8ToUTF16(title));
```

Comment:
We know `url` is a valid URL at this point?

Reply:
Sometimes `ntp_tiles::MostVisitedSites` handles invalid URLs gracefully, but I've added `is_valid` checks to be sure.